File: typography_migration_backup_20250830_221207/types/unified-profile.ts
/**
 * UNIFIED PROFILE INTERFACE
 * Single source of truth matching database schema exactly
 * Solves display_name and multiple interface conflicts
 */

export interface UnifiedProfile {
  user_id: string;
  email?: string;
  display_name?: string;
  full_name?: string;
  nickname?: string;
  avatar_url?: string;
  bio?: string;
  location?: string;
  website?: string;
  twitter_handle?: string;
  github_handle?: string;
  linkedin_handle?: string;
  created_at?: string;
  updated_at?: string;
  is_verified?: boolean;
  privacy_settings?: {
    show_email?: boolean;
    show_location?: boolean;
    show_social_links?: boolean;
  };
  preferences?: {
    theme?: 'light' | 'dark' | 'system';
    language?: string;
    notifications?: {
      email?: boolean;
      push?: boolean;
      in_app?: boolean;
    };
  };
  stats?: {
    total_matches?: number;
    wins?: number;
    losses?: number;
    win_rate?: number;
    current_streak?: number;
    best_streak?: number;
    total_points?: number;
    rank?: string;
    achievements?: string[];
  };
  
  // === LEGACY COMPATIBILITY ===
  id?: string;
  spa_points?: number;
  current_rank?: string;
  phone?: string;
  city?: string;
  district?: string;
  skill_level?: 'beginner' | 'intermediate' | 'advanced' | 'pro';
  verified_rank?: string;
  completion_percentage?: number;
  member_since?: string;
  role?: 'player' | 'club_owner' | 'both';
  active_role?: 'player' | 'club_owner';
  is_admin?: boolean;
  is_demo_user?: boolean;
  ban_status?: string;
  ban_reason?: string;
  banned_at?: string;
  banned_by?: string;
  elo?: number;
}

/**
 * DISPLAY NAME UTILITY
 * Matches database get_user_display_name() function exactly
 */
export function getDisplayName(profile: UnifiedProfile): string {
  // Use SAME logic as database function
  if (profile.display_name?.trim()) {
    return profile.display_name.trim();
  }
  if (profile.full_name?.trim()) {
    return profile.full_name.trim(); 
  }
  if (profile.nickname?.trim()) {
    return profile.nickname.trim();
  }
  if (profile.email?.trim()) {
    return profile.email.trim();
  }
  return `User ${profile.user_id.substring(0, 8)}`;
}

/**
 * PROFILE UPDATE DATA
 * Only fields that can be updated by user
 */
export interface ProfileUpdateData {
  full_name?: string;
  display_name?: string;
  phone?: string;
  bio?: string;
  avatar_url?: string;
  city?: string;
  district?: string;
  skill_level?: 'beginner' | 'intermediate' | 'advanced' | 'pro';
}

/**
 * PROFILE CREATION DATA
 * Minimal required data for creating new profile
 */
export interface ProfileCreateData {
  user_id: string;
  email?: string;
  current_rank?: string;
  spa_points?: number;
}

File: apps/sabo_flutter/lib/types/unified_profile.dart
// Types cho unified profile - port từ TypeScript
class UnifiedProfile {
  final String id;
  final String userId;
  final String? displayName;
  final String? phone;
  final String? bio;
  final String? skillLevel;
  final String? city;
  final String? district;
  final String? avatarUrl;
  final String? memberSince;
  final String role;
  final String activeRole;
  final String? verifiedRank;
  final String? email;
  final String? fullName;
  final String? currentRank;
  final int spaPoints;
  final String createdAt;
  final String updatedAt;
  final int completionPercentage;

  const UnifiedProfile({
    required this.id,
    required this.userId,
    this.displayName,
    this.phone,
    this.bio,
    this.skillLevel,
    this.city,
    this.district,
    this.avatarUrl,
    this.memberSince,
    required this.role,
    required this.activeRole,
    this.verifiedRank,
    this.email,
    this.fullName,
    this.currentRank,
    required this.spaPoints,
    required this.createdAt,
    required this.updatedAt,
    required this.completionPercentage,
  });

  UnifiedProfile copyWith({
    String? id,
    String? userId,
    String? displayName,
    String? phone,
    String? bio,
    String? skillLevel,
    String? city,
    String? district,
    String? avatarUrl,
    String? memberSince,
    String? role,
    String? activeRole,
    String? verifiedRank,
    String? email,
    String? fullName,
    String? currentRank,
    int? spaPoints,
    String? createdAt,
    String? updatedAt,
    int? completionPercentage,
  }) {
    return UnifiedProfile(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      displayName: displayName ?? this.displayName,
      phone: phone ?? this.phone,
      bio: bio ?? this.bio,
      skillLevel: skillLevel ?? this.skillLevel,
      city: city ?? this.city,
      district: district ?? this.district,
      avatarUrl: avatarUrl ?? this.avatarUrl,
      memberSince: memberSince ?? this.memberSince,
      role: role ?? this.role,
      activeRole: activeRole ?? this.activeRole,
      verifiedRank: verifiedRank ?? this.verifiedRank,
      email: email ?? this.email,
      fullName: fullName ?? this.fullName,
      currentRank: currentRank ?? this.currentRank,
      spaPoints: spaPoints ?? this.spaPoints,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      completionPercentage: completionPercentage ?? this.completionPercentage,
    );
  }
}

// Utility function để get display name giống web app
String getDisplayName(UnifiedProfile profile) {
  return profile.displayName ??
      profile.fullName ??
      profile.email?.split('@').first ??
      'Người chơi';
}

// Player stats cho profile
class PlayerStats {
  final int elo;
  final int spa;
  final int totalMatches;
  final double winRate;

  const PlayerStats({
    required this.elo,
    required this.spa,
    required this.totalMatches,
    required this.winRate,
  });
}

// Player ranking cho profile
class PlayerRanking {
  final int rankingPosition;
  final String tier;

  const PlayerRanking({
    required this.rankingPosition,
    required this.tier,
  });
}
